File: Problems/Algorithms/456. 132 Pattern/132Pattern.kt
/* The following solution using ArrayDeque will invoke
 * time limit error for the last test.
 * A workaround is to use Stack from Java...
 */
class Solution {
    fun find132pattern(nums: IntArray): Boolean {
        val stack = ArrayDeque<Int>()
        var third = Int.MIN_VALUE
        
        for (i in nums.size-1 downTo 0) {
            if (nums[i] < third) return true
            
            while (!stack.isEmpty() && stack.last() < nums[i]) {
                third = stack.removeLast()
            }
            stack.add(nums[i])
        }
        
        return false
    }
}

File: Problems/Algorithms/456. 132 Pattern/132-pattern.cpp
#include <stack>
#include <vector>

using namespace std;

class Solution {
public:
    bool find132pattern(vector<int>& nums) {
        stack<int> stack;
        int third = INT_MIN;

        for (int i = nums.size() - 1; i >= 0; i--) {
            if (nums[i] < third) {
                return true;
            }

            while (!stack.empty() && stack.top() < nums[i]) {
                third = stack.top();
                stack.pop();
            }
            stack.push(nums[i]);
        }

        return false;
    }
};
